app/src/main/kotlin/com/hippo/ehviewer/ktor/Apache5Extensions.kt

```diff
@@ -28,7 +43,49 @@ fun Apache5EngineConfig.configureClient() {
                         setIoThreadCount(1)
                     }.build(),
                 )
+                if (isAtLeastQ) {
+                    val resolver = DnsResolver.getInstance()
+                    val executor = Dispatchers.IO.asExecutor()
+                    setDnsResolver(
+                        object : ApacheDnsResolver {
+                            override fun resolve(host: String) = query(host, resolver, executor).toTypedArray()
+                            override fun resolveCanonicalHostname(host: String) = unreachable()
+                        },
+                    )
+                }
             }.build(),
         )
     }
 }
+
+@RequiresApi(Build.VERSION_CODES.Q)
+fun query(host: String, resolver: DnsResolver, executor: Executor) = runBlocking {
+    suspend fun doQuery(type: Int) = suspendCoroutine { cont ->
+        resolver.query(
+            null,
+            host,
+            type,
+            DnsResolver.FLAG_EMPTY,
+            executor,
+            null,
+            object : DnsResolver.Callback<List<InetAddress>> {
+                override fun onAnswer(result: List<InetAddress>, r: Int) {
+                    cont.resume(result)
+                }
+                override fun onError(e: DnsResolver.DnsException) {
+                    cont.resumeWithException(e)
+                }
+            },
+        )
+    }
+
+    try {
+        parZip(
+            { doQuery(DnsResolver.TYPE_A) },
+            { doQuery(DnsResolver.TYPE_AAAA) },
+            { a, b -> (a + b).shuffled() },
```

Comment:
Why not use the overload that queries with both IPv4 and IPv6? And shuffling will break rfc6724 sorting.